source/module_xc/xc_functional_gradcorr.cpp

```diff
@@ -94,17 +114,36 @@ void XC_Functional::gradcorr(double &etxc, double &vtxc, ModuleBase::matrix &v,
 
 	if(GlobalV::NSPIN == 4&&(GlobalV::DOMAG||GlobalV::DOMAG_Z))
 	{
-		ModuleBase::GlobalFunc::ZEROS(rhotmp1, GlobalC::rhopw->nrxx);
-		ModuleBase::GlobalFunc::ZEROS(rhogsum1, GlobalC::rhopw->npw);
 		rhotmp2 = new double[GlobalC::rhopw->nrxx];
-		ModuleBase::GlobalFunc::ZEROS(rhotmp2, GlobalC::rhopw->nrxx);
+		rhogsum2 = new std::complex<double>[GlobalC::rhopw->npw];
  		neg = new double [GlobalC::rhopw->nrxx];
-		ModuleBase::GlobalFunc::ZEROS(neg, GlobalC::rhopw->nrxx);
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static, 1024)
+#endif
+		for(int ir=0; ir<GlobalC::rhopw->nrxx; ir++)
+		{
+			rhotmp1[ir] = 0.0;
+			rhotmp2[ir] = 0.0;
+			neg[ir] = 0.0;
```

Comment:
IMHO the compiler has specific optimization for filling zeros. Does OpenMP helps here?

Reply:
Using a for loop to set zero of an array dose not trigger the optimization of the compiler, what ever for any -O flag

The optimization for filling zeros was implemented in `memset` of libc.
Passing zero to `memset` will trigger the lazy page allocation of the system, the memory page will actually on its first write, and it will be map to a zero page before.
Whether it works is depend on the libc and the system kernel

Reply:
> Using a for loop to set zero of an array dose not trigger the optimization of the compiler, no matter what -O flag
> 
> The optimization for filling zeros was implemented in `memset` of libc. If the memory block has not been initialized(just called `new` or `malloc` before `memset`), passing zero to `memset` will trigger the lazy page allocation of the system(alloc on write, and it will be cyclically mapped to a same readable zero page before first time of writing). Whether it works is depending on the libc and the system kernel

@Alcanderian You are partly right. I read a [passage](https://travisdowns.github.io/blog/2020/01/20/zero.html) about the optimization of compiler on this, and successfully replicated the result by https://godbolt.org/ . For `-O2` and above, the compiler will make filling zeros a call to `memset`. For GCC, you can see details in `-ftree-loop-distribute-patterns` in [optimization details](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html). Clang has a similar behavior. I have no ideas why this optimization is not documented.

I also tested on something like `for (auto i=0;i<10000;i++) a[i] = 0;b[i]=0;` : following the experiment I gave above, the compiler will make it two calls to `memset`. And if OpenMP for-loop is indicated, it will fall back to a normal for-loop.

However, I do not test the performance difference between them. I guess `memset` should have a better performance regarding OpenMP loops, for the bottleneck lies in the main memory bandwidth, and the number of threads might not help. This is my supposition; I might be wrong, and of course you can try it out yourself.

Reply:
> I think it is no need to fill zeros on these arrays, but I am not sure

@dyzheng Would you take a look on this? Or if you know someone can help?

Reply:
Maybe my compiler version is too old(GCC 7.5) and it does not trigger the optimization.
I measure the performance by timing in my test, and found cost time is only vanish when calling `memset` explicitly.

And accroding to the report three weeks ago, you can see that the main memory bandwidth peak of an application is depending on how many core dose it occupy.
An app only occupys a single core will only get 13-15GB/s bandwidth and can be up to 70-100GB/s if using multiple cores.
So if we could not take away the initialization, I highly recommend using multithreading

At last I will try to explain the performance of memset.
About memset
1. `memset` may use AVX2 depending on the libc compile option

About AVX2(SIMD)
1. SIMD memory operation is much fater than scalar operation when the data is located in some level of system cache(L1/L2), but no help when data is located in the main memory

Peformance
1. If the memory block size is smaller than L3 cache size, memset may faster than for loop
2. If the memory block size is smaller or equal with L1/L2 cache size, `memset` will be much faster, but the size of L1/L2 cache is too small(32KB and 512KB-1MB generally), and the bandwidth of L1/L2 cache is up to 100GB/s/core, so such a cost(about 1e-4 to 1e-5 second) of memory operation could be ignored
4. If the memory block size is much bigger than the system cache(3x larger), the perforamace of `memset` and for loop is almost the same

Reply:
> > I think it is no need to fill zeros on these arrays, but I am not sure
> 
> @dyzheng Would you take a look on this? Or if you know someone can help?

I think `memset` is good for this case.

Reply:
> > > I think it is no need to fill zeros on these arrays, but I am not sure
> > 
> > 
> > @dyzheng Would you take a look on this? Or if you know someone can help?
> 
> I think `memset` is good for this case.

@dyzheng I’m asking whether those array needs initialization.